File: rust/core/src/raw.rs
//! Raw binary framing for the surrogate nervous system (Cat-8 link layer).
//!
//! Mirrors `src/new_body/raw.py`. Frames are little-endian, magic-prefixed,
//! and protected by an IEEE 802.3 CRC32 that is byte-compatible with Python's
//! `zlib.crc32` (verified against the `123456789` test vector).

use serde::Serialize;

pub const MAGIC: &[u8; 4] = b"NB0\x01";
pub const VERSION: u8 = 1;

pub const PROTO_CONTROL: u8 = 0;
pub const PROTO_SENSORY: u8 = 1;
pub const PROTO_HAPTIC: u8 = 2;
pub const PROTO_THERMAL: u8 = 3;
pub const PROTO_UPLINK: u8 = 4;
pub const PROTO_BIOMETRIC: u8 = 5; // human body-system telemetry
pub const PROTO_COGNITIVE: u8 = 6; // mind / cognitive-module telemetry

const HEADER_SIZE: usize = 4 + 1 + 1 + 2 + 8 + 2; // magic..payload_len
const MIN_SIZE: usize = HEADER_SIZE + 4; // + crc32

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct Frame {
    pub protocol: u8,
    pub port: u16,
    pub timestamp_us: u64,
    pub payload: Vec<u8>,
}

/// IEEE 802.3 CRC32 (reflected, poly 0xEDB88320) — matches `zlib.crc32`.
pub fn crc32(data: &[u8]) -> u32 {
    let mut crc: u32 = 0xFFFF_FFFF;
    for &byte in data {
        crc ^= byte as u32;
        for _ in 0..8 {
            if crc & 1 != 0 {
                crc = (crc >> 1) ^ 0xEDB8_8320;
            } else {
                crc >>= 1;
            }
        }
    }
    !crc
}

impl Frame {
    pub fn encode(&self) -> Vec<u8> {
        let mut body = Vec::with_capacity(MIN_SIZE + self.payload.len());
        body.extend_from_slice(MAGIC);
        body.push(VERSION);
        body.push(self.protocol);
        body.extend_from_slice(&self.port.to_le_bytes());
        body.extend_from_slice(&self.timestamp_us.to_le_bytes());
        body.extend_from_slice(&(self.payload.len() as u16).to_le_bytes());
        body.extend_from_slice(&self.payload);
        let crc = crc32(&body);
        body.extend_from_slice(&crc.to_le_bytes());
        body
    }

    pub fn decode(data: &[u8]) -> Result<Frame, String> {
        if data.len() < MIN_SIZE {
            return Err("frame too short".into());
        }
        if &data[0..4] != MAGIC {
            return Err(format!("bad magic: {:?}", &data[0..4]));
        }
        if data[4] != VERSION {
            return Err(format!("unsupported version: {}", data[4]));
        }
        let protocol = data[5];
        let port = u16::from_le_bytes([data[6], data[7]]);
        let timestamp_us = u64::from_le_bytes(data[8..16].try_into().unwrap());
        let plen = u16::from_le_bytes([data[16], data[17]]) as usize;
        let body_end = HEADER_SIZE + plen;
        if data.len() < body_end + 4 {
            return Err("truncated payload".into());
        }
        let payload = data[HEADER_SIZE..body_end].to_vec();
        let crc = u32::from_le_bytes(data[body_end..body_end + 4].try_into().unwrap());
        let computed = crc32(&data[..body_end]);
        if computed != crc {
            return Err(format!("crc mismatch: computed {computed:#x} != {crc:#x}"));
        }
        Ok(Frame {
            protocol,
            port,
            timestamp_us,
            payload,
        })
    }

    pub fn size(&self) -> usize {
        self.encode().len()
    }
}

/// Decode a concatenated frame stream, returning decoded frames and any
/// trailing bytes that don't yet form a complete frame.
pub fn decode_stream(buffer: &[u8]) -> Result<(Vec<Frame>, Vec<u8>), String> {
    let mut frames = Vec::new();
    let mut offset = 0;
    let n = buffer.len();
    while offset + MIN_SIZE <= n {
        if &buffer[offset..offset + 4] != MAGIC {
            return Err(format!("bad magic at offset {offset}"));
        }
        let plen = u16::from_le_bytes([buffer[offset + 16], buffer[offset + 17]]) as usize;
        let end = offset + HEADER_SIZE + plen + 4;
        if n < end {
            break;
        }
        frames.push(Frame::decode(&buffer[offset..end])?);
        offset = end;
    }
    Ok((frames, buffer[offset..].to_vec()))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn sample() -> Frame {
        Frame {
            protocol: PROTO_SENSORY,
            port: 3,
            timestamp_us: 1_234_567,
            payload: vec![0xDE, 0xAD, 0xBE, 0xEF],
        }
    }

    #[test]
    fn roundtrip() {
        let frame = sample();
        let bytes = frame.encode();
        assert_eq!(Frame::decode(&bytes).unwrap(), frame);
    }

    #[test]
    fn crc_mismatch_detected() {
        let mut bytes = sample().encode();
        let last = bytes.len() - 1;
        bytes[last] ^= 0xFF;
        assert!(Frame::decode(&bytes).is_err());
    }

    #[test]
    fn bad_magic_rejected() {
        let mut bytes = sample().encode();
        bytes[0] = 0xFF;
        assert!(Frame::decode(&bytes).is_err());
    }

    #[test]
    fn truncated_payload_rejected() {
        let bytes = sample().encode();
        assert!(Frame::decode(&bytes[..bytes.len() - 2]).is_err());
    }

    #[test]
    fn stream_decode_and_leftover() {
        let a = sample();
        let b = Frame {
            port: 7,
            timestamp_us: 99,
            protocol: PROTO_SENSORY,
            payload: vec![0x01, 0x02],
        };
        let mut buf = a.encode();
        buf.extend_from_slice(&b.encode());
        buf.extend_from_slice(&[0x00, 0x01]); // trailing partial frame
        let (frames, leftover) = decode_stream(&buf).unwrap();
        assert_eq!(frames.len(), 2);
        assert_eq!(frames[0].port, 3);
        assert_eq!(frames[1].port, 7);
        assert_eq!(leftover, vec![0x00, 0x01]);
    }

    #[test]
    fn crc32_parity_with_standard_vector() {
        // CRC32("123456789") == 0xCBF43926 — must match Python's zlib.crc32.
        assert_eq!(crc32(b"123456789"), 0xCBF4_3926);
    }
}

File: src/new_body/raw.py
"""Raw binary framing for the surrogate nervous system (Cat-8 link layer).

Encodes sensory/control frames into a compact, checksummed byte layout that
mimics the sub-millisecond serialization described in the EDD section 2. Pure
stdlib (``struct`` + ``zlib``) — no external dependencies, and CRC32 is
byte-compatible with the Rust ``new-body-core`` implementation.

Frame layout (little-endian)::

    +--------+---------+----------+------+------+-------------+----------+---------+
    | magic  | version | protocol | port |  ts  | payload_len | payload |  crc32  |
    | 4s     | B       | B        | H    |  Q   | H           | ...      | I       |
    +--------+---------+----------+------+------+-------------+----------+---------+
"""

import struct
import zlib
from dataclasses import dataclass

MAGIC = b"NB0\x01"
VERSION = 1

# Protocol identifiers (subsystem link classes).
PROTO_CONTROL = 0
PROTO_SENSORY = 1
PROTO_HAPTIC = 2
PROTO_THERMAL = 3
PROTO_UPLINK = 4
PROTO_BIOMETRIC = 5  # human body-system telemetry
PROTO_COGNITIVE = 6  # mind / cognitive-module telemetry
PROTO_NATURE = 7  # human nature constructs (drives, values, moral, higher)

_HEADER = struct.Struct("<4sBBHQH")  # magic, version, protocol, port, ts, payload_len
_CRC = struct.Struct("<I")

# Minimum on-wire size: header + crc (no payload).
_MIN_SIZE = _HEADER.size + _CRC.size


@dataclass
class Frame:
    """A single Cat-8 link-layer frame."""

    protocol: int
    port: int
    timestamp_us: int
    payload: bytes

    def __bytes__(self) -> bytes:
        return encode_frame(self)

    @property
    def size(self) -> int:
        return len(encode_frame(self))


def encode_frame(frame: Frame) -> bytes:
    if not 0 <= frame.protocol <= 0xFF:
        raise ValueError("protocol out of range")
    if not 0 <= frame.port <= 0xFFFF:
        raise ValueError("port out of range")
    header = _HEADER.pack(
        MAGIC,
        VERSION,
        frame.protocol,
        frame.port,
        frame.timestamp_us,
        len(frame.payload),
    )
    body = header + frame.payload
    crc = zlib.crc32(body) & 0xFFFFFFFF
    return body + _CRC.pack(crc)


def decode_frame(data: bytes) -> Frame:
    if len(data) < _MIN_SIZE:
        raise ValueError("frame too short")
    magic, version, protocol, port, ts, plen = _HEADER.unpack_from(data, 0)
    if magic != MAGIC:
        raise ValueError(f"bad magic: {magic!r}")
    if version != VERSION:
        raise ValueError(f"unsupported version: {version}")
    expected = _HEADER.size + plen + _CRC.size
    if len(data) < expected:
        raise ValueError("truncated payload")
    payload = data[_HEADER.size : _HEADER.size + plen]
    (crc,) = _CRC.unpack_from(data, _HEADER.size + plen)
    body = data[: _HEADER.size + plen]
    if (zlib.crc32(body) & 0xFFFFFFFF) != crc:
        raise ValueError("crc mismatch")
    return Frame(protocol=protocol, port=port, timestamp_us=ts, payload=payload)


def decode_stream(buffer: bytes) -> tuple[list[Frame], bytes]:
    """Decode a concatenated frame stream.

    Returns the decoded frames and any trailing bytes that do not yet form a
    complete frame (re-feed them once more data arrives).
    """
    frames: list[Frame] = []
    offset = 0
    n = len(buffer)
    while offset + _MIN_SIZE <= n:
        magic, _version, _proto, _port, _ts, plen = _HEADER.unpack_from(buffer, offset)
        if magic != MAGIC:
            raise ValueError(f"bad magic at offset {offset}")
        end = offset + _HEADER.size + plen + _CRC.size
        if n < end:
            break  # need more bytes
        frames.append(decode_frame(buffer[offset:end]))
        offset = end
    return frames, buffer[offset:]
